### DISCUSSION TITLE
New Grails 6 CLI/FORGE: What it changed and what was deprecated/removed

### DISCUSSION THREAD

**Discussion by jamesfredley** [apache/grails-core#13562] [General]
With the New Grails 6 CLI/FORGE, material changes were made to new default applications in Grails, especially 6.2.x, and those changes do not seem to be well known or understood yet.   These changes unfortunately invalidate many historical stack overflow questions, blogs posts and other historical documentation in addition to portions of the currently published latest Grails documentation.  

**For a new default grails application, should the Grails Shell, Custom Scripts and Grails Profiles have been removed this quickly or should they have coexisted a bit longer with the new CLI/FORGE, until clearer paths were documented for all prior functionality, especially the new to grails basics?**  

It is still possible to use/add Grails Shell to a Grails 6 app and continue to use Custom Scripts, `grails create-* [DOMAIN_CLASS_NAME]`, `grails generate-* [DOMAIN_CLASS_NAME]` , `grails test package` and `grails dbm-gorm-diff <<filename>>.groovy --add` commands by adding the grails-shell dependency.  Versus the new gradle with GrailsApplicationCommand way `gradlew runCommand "-Pargs=generate-controller [DOMAIN_CLASS_NAME]|*`

`implementation "org.grails:grails-shell:6.1.2"`

The following was still needed in dependencies for my 6.2.0 app to run.  That doesn't align with profiles being removed.

`profile ("org.grails.profiles:web")`

https://github.com/grails/grails-core/issues/13399
https://grails.org/blog/2023-08-03-introducing-new-grails-forge.html
https://medium.com/@puneetbehl/a-comprehensive-guide-to-custom-application-command-in-grails-framework-f7496406ac50

"Since Grails 6, we stopped using/creating Grails profiles."  "The New Grails Forge also introduces Grails Profiles through application-types"

"While the revamped CLI offers enhanced performance, it operates entirely offline, lacking awareness of the Grails application or its plugins. This change disrupts the traditional approach to executing custom scripts within the Grails environment. Previously, developers could create custom scripts to automate tasks or extend functionality. However, with the CLI’s new offline mode, the concept of custom scripts becomes obsolete."

"With Grails 6, we're embracing a new era where the Grails CLI wrapper will no longer be included in new projects."

https://github.com/grails/grails-core/issues/13116
https://github.com/grails/grails-core/issues/13103 - environments are still supported, but the generated application.yml didn't include them as it has historically
https://github.com/grails/grails-core/issues/13512
https://github.com/grails/grails-core/issues/13499
https://github.com/grails/grails-core/issues/13108

**Comment by rainboyan** [comment]
@jamesfredley 

Thank you for collecting this useful information.

Here's a related discussion about the motivation behind the development of the new CLI/Forge.

> Ok from what I the Grails team at the time this was decided was just @puneetbehl with maybe some community help from a couple of people periodically. I can only imagine being the only full-time developer on a project as big and expansive as Grails. You have to plan features, implement features, fix bugs, evaluate/give feedback on pull requests, merge pull requests, deal with any consequences of those merges, update documentation, update plugins, do releases of Grails core and all of its underlying plugins, trying to get community members involved, etc. That is a lot to put on one person but Puneet did it and did it enthusiasitcly.

> Little over a year ago I think Puneet saw that I was answer questions on Slack and invited me to a weekly engineering meeting. When I first started attending there were a couple of other community members but as time went on, for quite a few meetings is was just Puneet and I talking about Grails. Again from what I was told and understood the old CLI and forge site was a pain to maintain and Puneet wanted to "upgrade" it to something new that would be easier for him to maintain, taking the ideas from the micronaut site and making it so everything was initiated through gradle rather than maintaining two ways of doing things the CLI and gradle.  Now I could be wrong but I thought the console and shell were still accisible  through Gradle commands.. Unfortuentely from this upgrade you lost the profiles and things like the Angular and Vue profiles and provided integration with those technologies.

> Towards the end of last year, Puneet finally started getting more help, which I though was good to see, but didn't know how short lived it would be. I know there was talk coming from @zacharyaklein about possibly bringing some of those profiles back under the new CLI, as he started joining the meeting periodically,  but I don't think he ever found the time to do that.

https://www.linen.dev/s/grails/t/18896892/i-m-a-noob-to-grails-so-sorry-for-the-basic-question-but-i-a

**Comment by jamesfredley** [comment]
More discussion on https://github.com/grails/grails-core/issues/13569

**Comment by codeconsole** [comment]
I think the term [deprecate](https://github.com/grails/grails-core/issues/13399) got lost in translation and ended up turning into the word [drop](https://github.com/grails/grails-gradle-plugin/pull/287)
